caliban_toolbox/dataset_builder.py

```diff
@@ -0,0 +1,571 @@
+
+        data_dict: dictionary of training data
+        resize: flag to control resizing of the data.
+            Valid arguments:
+                    - False. No resizing
+                    - by_tissue. Resizes by median cell size within each tissue type
+                    - by_image. Resizes by median cell size within each image
+        output_shape: output shape for image data
+        resize_target: desired median cell size after resizing
+        resize_tolerance: sets maximum allowable ratio between resize_target and
+            median cell size before resizing occurs
+        """
+        X, y = data_dict['X'], data_dict['y']
+        tissue_list = np.array(data_dict['tissue_list'])
+        platform_list = np.array(data_dict['platform_list'])
+
+        if not resize:
+            # no resizing
+            X_new, y_new = reshape_training_data(X_data=X, y_data=y, resize_ratio=1,
+                                                 final_size=output_shape, stride_ratio=1)
+
+            # to preserve category labels, we need to figure out how much the array grew by
+            multiplier = int(X_new.shape[0] / X.shape[0])
+
+            # then we duplicate the labels in place to match expanded array size
+            tissue_list_new = [item for item in tissue_list for _ in range(multiplier)]
+            platform_list_new = [item for item in platform_list for _ in range(multiplier)]
+
+        else:
+            X_new, y_new, tissue_list_new, platform_list_new = [], [], [], []
+
+            if resize == 'by_tissue':
+                batch_ids = np.unique(tissue_list)
+            else:
+                batch_ids = np.arange(0, X.shape[0])
+
+            # loop over each batch
+            for batch_id in batch_ids:
+
+                # get tissue types that match current tissue type
+                if isinstance(batch_id, str):
+                    batch_idx = np.isin(tissue_list, batch_id)
+
+                # get boolean index for current image
+                else:
+                    batch_idx = np.arange(X.shape[0]) == batch_id
+
+                X_batch, y_batch = X[batch_idx], y[batch_idx]
+                tissue_list_batch = tissue_list[batch_idx]
+                platform_list_batch = platform_list[batch_idx]
+
+                # compute appropriate resize ratio
+                median_cell_size = compute_cell_size({'X': X_batch, 'y': y_batch}, by_image=False)
+
+                # check for empty images
+                if median_cell_size is not None:
+                    resize_ratio = np.sqrt(resize_target / median_cell_size)
+                else:
+                    resize_ratio = 1
+
+                # resize the data
+                X_batch_resized, y_batch_resized = \
+                    reshape_training_data(X_data=X_batch, y_data=y_batch,
+                                          resize_ratio=resize_ratio, final_size=output_shape,
+                                          tolerance=resize_tolerance)
```

Comment:
No need for the `\` if the line breaks are inside a paranethesis.

```suggestion
                X_batch_resized, y_batch_resized = reshape_training_data(
                    X_data=X_batch, y_data=y_batch,
                    resize_ratio=resize_ratio, final_size=output_shape,
                    tolerance=resize_tolerance)
```